### ISSUE TITLE
Indicate and Display Details of GPG Signed Commits and Tags

### ISSUE THREAD

**Issue by nebhale** [ForkIssues/Tracker#18]
Recently, GitHub made it very easy to both create and view GPG signed commits.

This focus has caused [some projects](https://github.com/cloudfoundry/cf-java-client=) to adopt the signing of commits and tags as expected behaviors.

![screen shot 2017-04-28 at 09 45 59](https://cloud.githubusercontent.com/assets/60754/25538433/8bfa23ea-2bf7-11e7-8f61-9b51ffdfa88b.png)

![screen shot 2017-04-28 at 09 47 27](https://cloud.githubusercontent.com/assets/60754/25538483/bcb2723a-2bf7-11e7-806f-ef8cbd05dea4.png)

The Fork UI should expose both indicators that commits and tags have been signed as well as disclosures to view the signing information for those elements.

**Comment by DanPristupov** [comment]
I just checked the possibility to implement that.

1. To verify a commit's signature user needs to have the committer's public key on the computer. For example if you want GitHub to check **your** signatures you need to **upload** your public key to GitHub. But there's no way to get gpg keys locally, until you work in a team and share them. This makes 99% of signature verifications to return 'ERRSIG', which means 'missing key'.

2. GPG signature verification is a quite time consuming operation. GitHub can afford this because they store this data. Fork can't store verification information for each commit in every your repository. I can make the request lazily or  add signature verification icon to the revision details view, but (1) makes this almost pointless.

Please let me know if I'm missing something.

**Comment by nebhale** [comment]
To address number one, it's quite common to [upload GPG keys to shared key servers](http://hkps.pool.sks-keyservers.net/pks/lookup?search=0xb9c7ed5f135d02bc&fingerprint=on&op=index) which would give you a chance to attempt to retrieve keys and degrade gracefully if you couldn't find them (might be interesting to degrade the same way that GitHub does when presented with signed commits/tags that they don't have keys for).

On the second issue, I'd address it by validating signatures on view; i.e not in the table at the top, but only when you expanded the detail in the lower pane.  This would allow you to minimize the calculation on large repository (although maybe you could work on calculating and caching in the background while the application is idle since most people use the same repos over and over again).

Failing that, maybe since you've now got a GitHub integration, you could use that token and [GitHub APIs](https://developer.github.com/changes/2016-04-04-git-signing-api-preview/), letting GitHub do the heavy lifting and call it a feature only for GitHub repositories (unless BitBucket has some equivalent API).

**Comment by DanPristupov** [comment]
Thanks for the explanation. I don't have time for this right now, but I'll get back to it when I finish more important tasks and issues (interactive rebase and integration with git services).

**Comment by nebhale** [comment]
Yep, that's totally understandable.  Just wanted to get it on the list.

**Comment by infotexture** [comment]
If signature verification performance is the main blocker here, perhaps it would make sense to add a simple indicator that a commit has been signed _(with no performance penalty)_, and make signature verification a separate (manual) step that runs when users click the "signed" indicator/key ID.

**Comment by Turbo87** [comment]
@DanPristupov regarding the "where do the public keys come from" question: the "Verifying Commit Signatures with SSH" section of https://blog.dbrgn.ch/2021/11/16/git-ssh-signatures/ might give a bit of insight. git itself apparently supports a config setting pointing to a file with a list of trusted public (SSH) keys.
